nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py

```diff
@@ -0,0 +1,264 @@
+            version=dlm_software_version.version,
+            alias=dlm_software_version.alias,
+            release_date=dlm_software_version.release_date,
+            end_of_support_date=dlm_software_version.end_of_support,
+            documentation_url=dlm_software_version.documentation_url,
+            long_term_support=dlm_software_version.long_term_support,
+            pre_release=dlm_software_version.pre_release,
+            status=status_active,
+            _custom_field_data=dlm_software_version._custom_field_data,
+        )
+        core_software_version.save()
+
+        # Work around created field's auto_now_add behavior
+        CoreSoftwareVersion.objects.filter(id=core_software_version.id).update(created=dlm_software_version.created)
+        core_software_version.refresh_from_db()
+
+        # Migrate "Software on Device" relationships to the Device.software_version foreign key
+        for relationship_association in RelationshipAssociation.objects.filter(
+            relationship__key="device_soft",
+            source_type=core_software_version_ct,
+            source_id=core_software_version.id,
+            destination_type=device_ct,
+        ):
+            device = Device.objects.get(id=relationship_association.destination_id)
+            device.software_version = core_software_version
+            device.save()
+
+        # Migrate "Software on InventoryItem" relationships to the InventoryItem.software_version foreign key
+        for relationship_association in RelationshipAssociation.objects.filter(
+            relationship__key="inventory_item_soft",
+            source_id=core_software_version.id,
+            source_type=core_software_version_ct,
+            destination_type=inventory_item_ct,
+        ):
+            inventory_item = InventoryItem.objects.get(id=relationship_association.destination_id)
+            inventory_item.software_version = core_software_version
+            inventory_item.save()
+
+        # make tag manager available in migration for nautobot.core.models.utils.serialize_object
+        # https://github.com/jazzband/django-taggit/issues/101
+        # https://github.com/jazzband/django-taggit/issues/454
+        core_software_version.tags = TagsManager(
+            through=extras_models.TaggedItem,
+            model=CoreSoftwareVersion,
+            instance=core_software_version,
+            prefetch_cache_name="tags",
+        )
+
+        # Create an object change to document migration
+        ObjectChange.objects.create(
+            action=extras_choices.ObjectChangeActionChoices.ACTION_UPDATE,
+            change_context=extras_choices.ObjectChangeEventContextChoices.CONTEXT_ORM,
+            change_context_detail="Migrated from Nautobot App Device Lifecycle Management",
+            changed_object_id=core_software_version.id,
+            changed_object_type=core_software_version_ct,
+            object_data=serialize_object(core_software_version),
```

Comment:
no `object_data_v2`?

Reply:
I think we might have had problems doing the DRF serialization in migrations since we're not using the real models? I can double check to see if it looks like it will work here.

Reply:
Looks like I was skipping it in other migrations because we're not using the real models so the DRF serializer can't populate the fields properly. It looks like it's just the display and natural_slug fields that are affected here. Do you think we should use it as is?

```json
{
    "id": "4642a1f7-4bf1-49fc-ac75-66df0ddc02b7",
    "url": "/api/dcim/software-image-files/4642a1f7-4bf1-49fc-ac75-66df0ddc02b7/",
    "tags": [
        {
            "id": "a4b16991-191f-4940-95b4-1b5dd13e19f5",
            "url": "/api/extras/tags/a4b16991-191f-4940-95b4-1b5dd13e19f5/",
            "name": "Amber",
            "color": "ffc107",
            "created": "2023-11-09T16:00:00-08:00",
            "display": "Amber",
            "notes_url": "/api/extras/tags/a4b16991-191f-4940-95b4-1b5dd13e19f5/notes/",
            "description": "",
            "object_type": "extras.tag",
            "last_updated": "2023-11-10T06:59:44.936910-08:00",
            "natural_slug": "amber_a4b1",
            "custom_fields": {}
        }
    ],
    "status": {
        "id": "4e4c9239-7e70-4632-98aa-311090a61824",
        "url": "/api/extras/statuses/4e4c9239-7e70-4632-98aa-311090a61824/",
        "name": "Active",
        "color": "4caf50",
        "created": "2023-11-09T16:00:00-08:00",
        "display": "Status object (4e4c9239-7e70-4632-98aa-311090a61824)",
        "notes_url": "/api/extras/statuses/4e4c9239-7e70-4632-98aa-311090a61824/notes/",
        "description": "Unit is active",
        "object_type": "extras.status",
        "last_updated": "2023-11-10T06:57:23.366251-08:00",
        "natural_slug": "unknown",
        "custom_fields": {}
    },
    "created": "2024-02-20T15:18:46.750295-08:00",
    "display": "SoftwareImageFile object (4642a1f7-4bf1-49fc-ac75-66df0ddc02b7)",
    "notes_url": "/api/dcim/software-image-files/4642a1f7-4bf1-49fc-ac75-66df0ddc02b7/notes/",
    "object_type": "dcim.softwareimagefile",
    "download_url": "",
    "last_updated": "2024-02-21T12:08:43.505074-08:00",
    "natural_slug": "unknown",
    "custom_fields": {
        "custom_field_for_dlm_software": "hp_object_tags"
    },
    "default_image": false,
    "image_file_name": "test_hp_object_tags.bin",
    "image_file_size": null,
    "software_version": {
        "id": "13cd62e1-21e6-4ccc-ac6d-47e89cf6077a",
        "url": "/api/dcim/software-versions/13cd62e1-21e6-4ccc-ac6d-47e89cf6077a/",
        "alias": "device types",
        "status": {
            "id": "4e4c9239-7e70-4632-98aa-311090a61824",
            "url": "/api/extras/statuses/4e4c9239-7e70-4632-98aa-311090a61824/",
            "object_type": "extras.status"
        },
        "created": "2024-02-20T15:00:11.543075-08:00",
        "display": "SoftwareVersion object (13cd62e1-21e6-4ccc-ac6d-47e89cf6077a)",
        "version": "software version device types",
        "platform": {
            "id": "77ec06de-68a6-4b50-a4b3-24283ea111e3",
            "url": "/api/dcim/platforms/77ec06de-68a6-4b50-a4b3-24283ea111e3/",
            "object_type": "dcim.platform"
        },
        "notes_url": "/api/dcim/software-versions/13cd62e1-21e6-4ccc-ac6d-47e89cf6077a/notes/",
        "object_type": "dcim.softwareversion",
        "pre_release": false,
        "last_updated": "2024-02-21T12:08:43.402080-08:00",
        "natural_slug": "unknown",
        "release_date": null,
        "custom_fields": {
            "custom_field_for_dlm_software": "custom field for software version device types"
        },
        "documentation_url": "",
        "long_term_support": false,
        "end_of_support_date": null
    },
    "hashing_algorithm": "",
    "image_file_checksum": ""
}
```

Reply:
natural_slug and display are pretty irrelevant for change logging, I think that's close enough.